sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala

```diff
@@ -35,184 +34,337 @@ import org.apache.spark.util.Utils
+        parser.getCurrentToken match {
+          case VALUE_STRING =>
+            val stringValue = parser.getText
+            if (stringValue.contains("-")) {
+              // The format of this string will probably be "yyyy-mm-dd".
+              DateTimeUtils.millisToDays(DateTimeUtils.stringToTime(parser.getText).getTime)
+            } else {
+              // In Spark 1.5.0, we store the data as number of days since epoch in string.
+              // So, we just convert it to Int.
+              stringValue.toInt
+            }
 
-      case (START_ARRAY, ArrayType(st, _)) =>
-        convertArray(factory, parser, st)
+          case _ => failedConversion(parser, dataType)
+        }
+      }
 
-      case (START_OBJECT, MapType(StringType, kt, _)) =>
-        convertMap(factory, parser, kt)
+    case BinaryType =>
+      (parser: JsonParser) => convertField(parser) {
+        parser.getCurrentToken match {
+          case VALUE_STRING => parser.getBinaryValue
+          case _ => failedConversion(parser, dataType)
+        }
+      }
+
+    case dt: DecimalType =>
+      (parser: JsonParser) => convertField(parser) {
+        parser.getCurrentToken match {
+          case (VALUE_NUMBER_INT | VALUE_NUMBER_FLOAT) =>
+            Decimal(parser.getDecimalValue, dt.precision, dt.scale)
+
+          case _ => failedConversion(parser, dt)
+        }
+      }
+
+    case st: StructType =>
+      val fieldConverters = st.map(_.dataType).map(makeConverter)
+      (parser: JsonParser) => convertField(parser) {
+        parser.getCurrentToken match {
+          case START_OBJECT => convertObject(parser, st, fieldConverters)
+          case _ => failedConversion(parser, st)
+        }
+      }
+
+    case at: ArrayType =>
+      val elementConverter = makeConverter(at.elementType)
+      (parser: JsonParser) => convertField(parser) {
+        parser.getCurrentToken match {
+          case START_ARRAY => convertArray(parser, elementConverter)
+          case _ => failedConversion(parser, at)
+        }
+      }
+
+    case mt: MapType =>
+      val valueConverter = makeConverter(mt.valueType)
+      (parser: JsonParser) => convertField(parser) {
+        parser.getCurrentToken match {
+          case START_OBJECT => convertMap(parser, valueConverter)
+          case _ => failedConversion(parser, mt)
+        }
+      }
+
+    case udt: UserDefinedType[_] =>
+      makeConverter(udt.sqlType)
+
+    case _ =>
+      (parser: JsonParser) =>
+        failedConversion(parser, dataType)
+  }
+
+  /**
+   * This converts a field. If this is called after `START_OBJECT`, then, the next token can be
+   * `FIELD_NAME`. Since the names are kept in `JacksonParser.convertObject`, this `FIELD_NAME`
+   * token can be skipped as below. When this is called after `START_ARRAY`, the tokens become
+   * ones about values until `END_ARRAY`. In this case, we don't have to skip.
+   */
+  private def convertField(parser: JsonParser)(f: => Any): Any = {
```

Comment:
Sure, I will merge both `convertValue` and `convertField` but additional checking should be added for merging  https://github.com/apache/spark/pull/14102/files#r74034675 to be called ahead, because in `StringType`, it does not call `failedConversion` when it fails. For `StringType`, it should produce empty string when `parser.getTextLength < 1`.

This is also partly because I guess the failed cases are relatively low.